### PR TITLE
feat: add firebase cloud messaging

### PR DESCRIPTION
**Pull request by qu0cquyen**
## Status

**READY 

## Description

Add Firebase Cloud Messaging

## Type of Change

- [X] ✨ New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: packages/robusta_firebase_cloud_messaging/example/test/main_test.dart [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import 'package:flutter_test/flutter_test.dart';
+import 'package:robusta_firebase_cloud_messaging_example/main.dart';
+
+void main() {
+  group('test app', () {
+    testWidgets('can run app', (tester) async {
+      await expectLater(runner.run, returnsNormally);
```
Comment (vuongxuongminh) [inline packages/robusta_firebase_cloud_messaging/example/test/main_test.dart]: It should be `await expectLater(runner.run(), completes);`

File: packages/robusta_firebase_cloud_messaging/lib/src/extension.dart [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+      alert: _settings._alert,
+      announcement: _settings._announcement,
+      badge: _settings._badge,
+      carPlay: _settings._carPlay,
+      criticalAlert: _settings._criticalAlert,
+      provisional: _settings._provisional,
+      sound: _settings._sound,
+    );
+
+    log('User granted permission: $settings');
```
Comment (vuongxuongminh) [inline packages/robusta_firebase_cloud_messaging/lib/src/extension.dart]: It should be `container.read(loggerProvider).i('User granted permission: $settings');`

File: packages/robusta_firebase_cloud_messaging/lib/src/extension.dart [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    if (_notiRequestStrategy == NotiRequestStrategy.init) {
+      await permissionRequest(container);
+    }
+
+    final messaging = FirebaseMessaging.instance;
+
+    final notiAuthStatus = await messaging.getNotificationSettings();
+
+    if (notiAuthStatus.authorizationStatus == AuthorizationStatus.authorized) {
+      FirebaseMessaging.onMessage.listen((RemoteMessage rMessage) async {
```
Comment (vuongxuongminh) [inline packages/robusta_firebase_cloud_messaging/lib/src/extension.dart]: How about in-app messages and background messages?

File: packages/robusta_firebase_cloud_messaging/lib/src/extension.dart [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+/// {@template robusta_firebase_cloud_messaging}
+/// A Helper class for Firebase Cloud Messaiging
+/// {@endtemplate}
+class FirebaseCloudMessagingExtension implements DependenceExtension {
+  /// {@macro robusta_firebase_cloud_messaging}
+  FirebaseCloudMessagingExtension({
+    required EventManager eventManager,
+    NotiPermissionRequestSettings? settings,
+    NotiRequestStrategy? notiRequestStrategy,
```
Comment (vuongxuongminh) [inline packages/robusta_firebase_cloud_messaging/lib/src/extension.dart]: In this context permission request only notification, so prefix with `Noti` is redundance, WDYT?

File: packages/robusta_firebase_cloud_messaging/pubspec.yaml [hunk 5]
```diff
@@ -0,0 +1,23 @@
+name: robusta_firebase_cloud_messaging
+homepage: https://robusta.covalab.io
+repository: https://github.com/covalab/robusta
+issue_tracker: https://github.com/covalab/robusta/issues
+description: Providing bridge to integrate Robusta with Firebase Cloud Messaging features.
+
```
Comment (vuongxuongminh) [inline packages/robusta_firebase_cloud_messaging/pubspec.yaml]: missing `version`.

File: packages/robusta_firebase_cloud_messaging/lib/src/extension.dart [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+part './permission_settings.dart';
+
+/// {@template robusta_firebase_cloud_messaging}
+/// A Helper class for Firebase Cloud Messaiging
+/// {@endtemplate}
+class FirebaseCloudMessagingExtension implements DependenceExtension {
+  /// {@macro robusta_firebase_cloud_messaging}
+  FirebaseCloudMessagingExtension({
+    required EventManager eventManager,
```
Comment (vuongxuongminh) [inline packages/robusta_firebase_cloud_messaging/lib/src/extension.dart]: `EventManager` should be optional and fallback to event manager had set on `runner` if it's not set.

Things about if we use another event manager, how user can add event listeners to it?

**Review by vuongxuongminh — CHANGES_REQUESTED** [review]

File: packages/robusta_firebase_cloud_messaging/example/test/main_test.dart (hunk 1 above)
Comment (qu0cquyen) [reply packages/robusta_firebase_cloud_messaging/example/test/main_test.dart]: Done.

File: packages/robusta_firebase_cloud_messaging/lib/src/extension.dart (hunk 2 above)
Comment (qu0cquyen) [reply packages/robusta_firebase_cloud_messaging/lib/src/extension.dart]: Done.

File: packages/robusta_firebase_cloud_messaging/lib/src/extension.dart (hunk 3 above)
Comment (qu0cquyen) [reply packages/robusta_firebase_cloud_messaging/lib/src/extension.dart]: In-app Message should be in a separate extension.

File: packages/robusta_firebase_cloud_messaging/lib/src/extension.dart (hunk 4 above)
Comment (qu0cquyen) [reply packages/robusta_firebase_cloud_messaging/lib/src/extension.dart]: Done.

File: packages/robusta_firebase_cloud_messaging/pubspec.yaml (hunk 5 above)
Comment (qu0cquyen) [reply packages/robusta_firebase_cloud_messaging/pubspec.yaml]: Added

File: packages/robusta_firebase_cloud_messaging/lib/src/extension.dart (hunk 6 above)
Comment (qu0cquyen) [reply packages/robusta_firebase_cloud_messaging/lib/src/extension.dart]: Changed to

```dart
FirebaseCloudMessagingExtension({
    EventManager? eventManager,
    PermissionRequestSettings? settings,
    RequestStrategy? requestStrategy,
  })  : _eventManager = eventManager ?? DefaultEventManager(),
        _notiRequestStrategy = requestStrategy ?? RequestStrategy.init,
```

WDYT?

Comment (vuongxuongminh) [reply packages/robusta_firebase_cloud_messaging/lib/src/extension.dart]: `RequestStrategy` should be named `PermissionRequestStrategy`. 

`_eventManager` accepts null and fallback to event manager of runner if it's null.

**Review by vuongxuongminh — APPROVED** [review]

File: packages/robusta_firebase_messaging/lib/src/extension.dart [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+      container.read(tokenProvider.notifier).state = token;
+    });
+
+    /// This will get [Firebase Token] for the 1st boot
+    /// Since token won't change unless it expires
+    final token = await _messaging.getToken();
+
+    container.read(tokenProvider.notifier).state = token;
+    container.read(loggerProvider).d('Firebase Messaging Token: $token');
+
```
Comment (vuongxuongminh) [inline packages/robusta_firebase_messaging/lib/src/extension.dart]: Get Firebase token should be wrap on post frame callback too for handling exceptions on UI instead froze app on error.

**Review by vuongxuongminh — CHANGES_REQUESTED** [review]

**Review by vuongxuongminh — APPROVED** [review]

**Review by vuongxuongminh — APPROVED** [review]

**Comment by vuongxuongminh** [comment]
Merged! Thank you @qu0cquyen

**Pull request merged**